plugins/usi-dock/fu-usi-dock-mcu-device.c

```diff
@@ -0,0 +1,777 @@
+/*
+ * Copyright (C) 2021 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include "fu-usi-dock-child-device.h"
+#include "fu-usi-dock-common.h"
+#include "fu-usi-dock-dmc-device.h"
+#include "fu-usi-dock-firmware.h"
+#include "fu-usi-dock-mcu-device.h"
+
+struct _FuUsiDockMcuDevice {
+	FuHidDevice parent_instance;
+};
+
+G_DEFINE_TYPE(FuUsiDockMcuDevice, fu_usi_dock_mcu_device, FU_TYPE_HID_DEVICE)
+
+#define FU_USI_DOCK_MCU_DEVICE_TIMEOUT 5000 /* ms */
+
+/**
+ * FU_USI_DOCK_DEVICE_FLAG_VERFMT_HP:
+ *
+ * All version numbers should be stored in HP format.
+ *
+ * Since: 1.7.3
+ */
+#define FU_USI_DOCK_DEVICE_FLAG_VERFMT_HP (1 << 0)
+
+/* returned TRUE - got a expected report*/
+static gboolean
+fu_usi_dock_rx_filter(guint8 cmd, const guint8 *buf)
+{
+	if (cmd != buf[1])
+		return TRUE;
+
+	switch (cmd) {
+	case USBUID_ISP_DEVICE_CMD_MCU_JUMP2BOOT:
+		g_debug("got correct jump");
+		break;
+	case USBUID_ISP_INTERNAL_FW_CMD_TARGET_CHECKSUM:
+		if (buf[6] == FIRMWARE_IDX_AUDIO) {
+			g_debug("got a quick jump at audio updates");
+		} else {
+			g_debug("got a ignored report");
+			return FALSE;
+		}
+		break;
+	default:
+		break;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_usi_dock_mcu_device_tx(FuUsiDockMcuDevice *self,
+			  guint8 tag2,
+			  const guint8 *inbuf,
+			  gsize inbufsz,
+			  GError **error)
+{
+	guint8 buf[64] = {USB_HID_REPORT_ID2,
```

Comment:
I wonder if this buffer should be a struct that you define?  It seems cmd is always the first, and then it's a certain number of args.  Not sure if those are variable behaviors per command or same for all.